File: backend/app/api/routes/imports.py
import json

from fastapi import APIRouter, File, Form, UploadFile

from app.api.deps import SessionDep
from app.schemas.imports import ColumnMapping, ImportCommitResponse, ImportPreviewResponse, ImportRow
from app.services import import_service

router = APIRouter(prefix="/import", tags=["import"])


@router.post("/preview", response_model=ImportPreviewResponse)
async def preview_import(
    session: SessionDep,
    file: UploadFile = File(...),
    column_mapping: str = Form(...),
):
    file_bytes = await file.read()
    mapping = ColumnMapping.model_validate(json.loads(column_mapping))
    return import_service.preview(session, file_bytes, mapping)


@router.post("/commit", response_model=ImportCommitResponse)
def commit_import(rows: list[ImportRow], session: SessionDep):
    return import_service.commit(session, rows)

File: frontend/src/api/imports.ts
import { post, postForm } from "./client";
import type { ColumnMapping, ImportCommitResponse, ImportPreviewResponse, ImportRow } from "../types";

export async function previewImport(
  file: File,
  mapping: ColumnMapping
): Promise<ImportPreviewResponse> {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("column_mapping", JSON.stringify(mapping));
  return postForm<ImportPreviewResponse>("/import/preview", formData);
}

export const commitImport = (rows: ImportRow[]) =>
  post<ImportCommitResponse>("/import/commit", rows);
